actors/builtin/market/market_actor.go

```diff
@@ -156,22 +157,30 @@ func (a Actor) PublishStorageDeals(rt Runtime, params *PublishStorageDealsParams
 		proposals := AsDealProposalArray(adt.AsStore(rt), st.Proposals)
 		dbp := AsSetMultimap(adt.AsStore(rt), st.DealIDsByParty)
 		// All storage proposals will be added in an atomic transaction; this operation will be unrolled if any of them fails.
+
+		_, worker := builtin.RequestMinerControlAddrs(rt, params.Deals[0].Proposal.Provider)
+
 		for _, deal := range params.Deals {
-			if deal.Proposal.Provider != rt.Message().Caller() {
+
+			if worker != rt.Message().Caller() {
```

Comment:
Pull this check out of the loop over deals, but add a check that the deal provider is the same as the first deal's provider.

Note: we don't resolve the provider address yet, but maybe should? It's the actor's address, so could conceivably be the reorg-stable actor address instead of ID address. We should either resolve it or assert it's an ID address (the former is probably more helpful).

If you resolve it `s/provider/resolved provider/g` in the first sentence of this comment.